Major compaction: flush commitlog by forcing new active segment and flushing all tables

bhalevy opened the pull request:
Major compaction already flushes each table to make
sure it considers any mutations that are present in the
memtable for the purpose of tombstone purging.
See 64ec1c6ec64291064678e3818e3ffdee4294dbfe

However, tombstone purging may be inhibited by data
in commitlog segments based on `gc_time_min` in the
`tombstone_gc_state` (See f42eb4d1ce321e93b934ce23e128af29d23d66f8).

Flushing all sstables in the database release
all references to commitlog segments and there
it maximizes the potential for tombstone purging,
which is typically the reason for running major compaction.

However, flushing all tables too frequently might
result in tiny sstables.  Since when flushing all
keyspaces using `nodetool flush` the `force_keyspace_compaction`
api is invoked for keyspace successively, we need a mechanism
to prevent too frequent flushes by major compaction.

Hence a `compaction_flush_all_tables_before_major_seconds` interval
configuration option is added (defaults to 24 hours).

In the case that not all tables are flushed prior
to major compaction, we revert to the old behavior of
flushing each table in the keyspace before major-compacting it.

Fixes scylladb/scylladb#15777

elcallio reviewed (approved):
If we are ok with the additional overhead, this is a neat way to more or less ensure behaviour akin to before.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - Build
:x: - Sanity Tests
:white_check_mark: - Unit Tests
#### Failed Tests (1/21010):
- [test_lcs_table_promotion_major_compaction](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4349/testReport/junit/compaction_additional_test/TestLCSSSTablePromotion/Tests___Sanity_Tests___test_lcs_table_promotion_major_compaction) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_lcs_table_promotion_major_compaction)
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4349/
- Duration: 7 hr 33 min
- Builder: i-026e35dd13dbbd224

avikivity commented on compaction/task_manager_module.cc:
```diff
@@ -266,6 +266,7 @@ future<tasks::task_manager::task::progress> compaction_task_impl::get_progress(c
 }
 
 future<> major_keyspace_compaction_task_impl::run() {
+    co_await replica::database::flush_on_all_shards(_db);
```
What prevents a mutation to be added to commitlog right after this call?

Is the problem one of correctness (a tombstone will be incorrectly purged) or efficiency (compaction will detect the problem and avoid purging the tombstone)?

Flushing all tables can be a bad tradeoff in a cluster with many tables.

elcallio replied:
Nothing is failing. But since we now take potential data in commit log into account when deciding what time boundaries exist for GC, tests that assume they can add data, wait X time, compact, and voila, GC has happened won't always work, since unless commitlog is properly emptied as well, compaction might not (yet) be able to apply tombstones etc. 

Tests can handle this simply by calling nodetool flush instead, this will do the same as is done here in this patch (i.e. clear out active segment so memtable flush can empty all segments in a test). I modified a bunch of tests, and sent a patch that did the same with the test mentioned in the issue. 

This PR does the same thing unconditionally before every (major) compaction instead.

avikivity replied:
I see, so it was really a test problem.

Now, we need to consider if a flush of all tables is worth the additional tombstones collected, yes?

elcallio replied:
Yes. I am on the fence. But I agree with the author that it might be worth it to maintain behaviour for tests. Major compaction should not be that frequent, should it?
The other option is to patch tests that make assumptions. 

If the major compaction task could decide which CF:s it intends to work on before starting, we could also limit the flushes and CL segment discards of course. Perhaps a later improvement.

bhalevy replied:
> I see, so it was really a test problem.

A test is an application.

> Nothing is failing. But since we now take potential data in commit log into account when deciding what time boundaries exist for GC, tests that assume they can add data, wait X time, compact, and voila, GC has happened won't always work, since unless commitlog is properly emptied as well, compaction might not (yet) be able to apply tombstones etc.

Compaction purges tombstones if they meet 2 conditions:
1. They are eligible for purging based on the tombstone_gc mode (either gc_grace_seconds passed since the tombstone was written, or a successful repair included them in tombstone_gc_mode=repair)
2. The tombstones shadow no data in other sstables, based on bloom filter matching, and timestamp ranges of the matched sstables.

The reason tables are flushed in 64ec1c6ec64291064678e3818e3ffdee4294dbfe (#9514) is the 2nd point above - since mutations in the memtable are not present in sstables we flush the memtables to use the sstables shadowing logic.

This change adds to that "flushing" of the commitlog so to ensure the commitlog won't resurrect data (or tombstones) after tombstone_gc.

To avoid it, we'll to maintain some metadata in core like a bloom filter for the commitlog (can be done per segment) and min/max timestamp again for each segment, so that compaction will postpone tombstone purging until it can verify that the tombstones do not potentially shadow any data in the commitlog.

We can do that as a follow-up as an optimization to this patch.
Ans a similar approach can be taken for the memtable to save their flushing.

A big advantage of the latter approach is that it will work for tombstone_gc also in regular compaction.

elcallio replied:
No, this change does nothing to handle CL resurrection. That was handled in #15060, which ensures we don't try to GC things if there are potential CL entries that would collide with said GC. What this patch does it try to eliminate remaining CL data so that a compaction can be fully realized earlier. Equal results would been had from, for example, writing a bunch of unrelated data and thus similarly switching active CL segment. 

The "problem" with #15060, for tests and others that assume that only time waited affects GC-ability, is that whatever writes/deletes etc done by the test/program up until the waiting will end up with a portion of this data in a CL segment that is still active (allocating), and thus not eligible for deletion, even though memtable flush might clear out data for it. 
As a workaround explicit nodetool flush (the magic wand of all tests) nowadays forces rotating active CL segment as well. 

The above change here basically adds a nodetool flush to all major compactions.

bhalevy replied:
I see, so the corrected system behavior broke the test.  That makes sense.

> Now, we need to consider if a flush of all tables is worth the additional tombstones collected, yes?

Yes, I think it's worth it.
Major compaction is a rare event, and it is used many times by FEs and customers as a way to purge tombstone when too many expired tombstone were collected and regular compaction is not able to purge them for some reason, so this patch will make that step more complete.
We've seen cases in the field were lightly written tables pinned commitlog segments for days if not weeks and that might prevent tombstone gc to do its thing.
Requiring the user or test developer to flush all tables prior to major compaction is not-trivial and prone to human error, hence the automation in this patch should be preferred.

avikivity replied:
Ok. I agree.

But we need some test here.

bhalevy replied:
The `cleanup_tet.py::TestCleanup::test_cluster_restore_no_resurrection` dtest consistently fails without this patch and consistently passes with it.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4449/
- Duration: 2 hr 19 min
- Builder: spider2.cloudius-systems.com

bhalevy commented:
@avikivity ok to merge?

raphaelsc commented:
Just a note: one of the side affects with early flush is increased write amp, and the impact of course depends on frequency of majors.

elcallio commented:
I have no idea if a minor compaction is allowed to GC tombstones... :-/
The original patch (https://github.com/scylladb/scylladb/commit/f42eb4d1ce321e93b934ce23e128af29d23d66f8) will lock out any compaction where CL could cause resurrection. 
The goal here is to restore/preserve GC behaviour, mainly for tests. Another approach (as previously discussed) is forcing tests to do manual nodetool flush (which nowadays will also try to purge CL).

avikivity commented:
minor compaction will gc tombstones, just not those that potentially cover data in commitlog segments pinned by other tables

raphaelsc commented:
Do we have a concept of TTL for active segments? I am really worried about the new logic backfiring with tables, that is either very low or very high in write activity co-existing. Then they could share CL segments, and tables with high write activity could suffer by not being able to GC tombstones until tables with very low writes finally release the segments. I also wonder if the simplest solution was indeed the best we had on table. A tombstone-like record could also work while not introducing this cross-table dependency for tombstone GC efficiency. But maybe the problem I mentioned above is already solved, then my concern is not valid.

elcallio commented:
@raphaelsc : Segments don't have any concept of TTL as such, but the previous patch added keeping track of lowest mutation timestamp per segment, and retrieving it on compaction. Indeed, segments kept alive by tables not being memtable flushed could prevent GC thusly, which in a sense is an argument for the above change here, forcing global flushes etc on full comapction. Btw, perhaps you mean tablets, not tables, since I've previously suggesting keeping CL per tablet, not CF in that world... 

Not sure if https://github.com/scylladb/scylladb/commit/f42eb4d1ce321e93b934ce23e128af29d23d66f8 is the _easiest_ solution, but it probably is the safest/lowest overall impact one. I.e. does not add all that much work, though again, in pathological cases it can prevent GC somewhat...

What do you mean by "tombstone-like record"?

Note that we could alleviate the problem by working on making keeping track of sstable RP:s better/actually working, but any compaction across shards prevents that. Do we ever compact sstables created during a nodes current run (uptime) across shards? If not, just tracking highest RP originating in the current node run into resulting sstable(s) would allow us to effectively 
eliminate any segments from replay based on sstable set.

avikivity commented:
> Do we have a concept of TTL for active segments? I am really worried about the new logic backfiring with tables, that is either very low or very high in write activity co-existing.

If there are tables with high write activity then they will recycle commitlog quickly.

We have amount 8GB of commitlog per shard in average configurations. AT 2MB/s per shard commitlog write rate (relatively low) we recycle all of commitlog in 4000s = 1h. That's much lower than reasonable repair intervals or gc_grace_seconds.


> Then they could share CL segments, and tables with high write activity could suffer by not being able to GC tombstones until tables with very low writes finally release the segments. I also wonder if the simplest solution was indeed the best we had on table. A tombstone-like record could also work while not introducing this cross-table dependency for tombstone GC efficiency. But maybe the problem I mentioned above is already solved, then my concern is not valid.

bhalevy commented:
> > Do we have a concept of TTL for active segments? I am really worried about the new logic backfiring with tables, that is either very low or very high in write activity co-existing.
> 
> If there are tables with high write activity then they will recycle commitlog quickly.

But the slow tables will still hold up CL segments from recycling.

Maybe we need a TTL on the memtables instead so they are flushed after an idle timeout (e.g. 1 hour) even if the table has infrequent writes.

raphaelsc commented:
> > > Do we have a concept of TTL for active segments? I am really worried about the new logic backfiring with tables, that is either very low or very high in write activity co-existing.
> > 
> > 
> > If there are tables with high write activity then they will recycle commitlog quickly.
> 
> But the slow tables will still hold up CL segments from recycling.
> 
> Maybe we need a TTL on the memtables instead so they are flushed after an idle timeout (e.g. 1 hour) even if the table has infrequent writes.

Exactly. We can't assume write frequency. There are tables that are almost idle and could affect other co-existing tables that are tombstone heavy. One of the things I thought yesterday, later, was indeed flushing stale memtables. We have to handle it or otherwise users, with this scenario, might potentially face a regression.

bhalevy commented:
With https://github.com/scylladb/scylladb/issues/15920 covering the regular-compaction path, I think we can merge this PR as is.

bhalevy commented:
@avikivity, @denesb ok to merge this PR?

denesb commented:
> @avikivity, @denesb ok to merge this PR?

Fine by me.

mykaul commented:
@avikivity  - please review.

raphaelsc commented on compaction/task_manager_module.cc:
```diff
@@ -266,6 +266,7 @@ future<tasks::task_manager::task::progress> compaction_task_impl::get_progress(c
 }
 
 future<> major_keyspace_compaction_task_impl::run() {
+    co_await replica::database::flush_on_all_shards(_db);
```
But let's not block this series by this comment.

bhalevy replied:
It's a good point.
I need to rebase anyway, so I'll give it a shot.

raphaelsc replied:
we'd need a wrapper in database like: database::compact_all_sstables(table& t);

bhalevy replied:
@raphaelsc flushing all tables needs to happen once per major compaction while the call to compact_all_sstables is too low in the stack (top-level -> shard -> table) so you don't want to flush all tables per table.  We can do it once per shard, but it will still be in the task_manager module (shard_major_keyspace_compaction_task_impl)

raphaelsc replied:
If we're flushing multiple (say hundred) keyspaces, one after the other, it could be a problem that we're producing many small sstables, right?

raphaelsc replied:
for every keyspace compacted, we're flushing every memtable across all keyspaces and shards. early flushes contribute not only with additional compaction work but also write amplification.

raphaelsc replied:
the write amplifications comes from that fact that you're adding a small base sstable size, meaning more size tiers for the data to travel through.  that's somehow mitigated with min_sstable_size, but it's still additional write amp.

raphaelsc replied:
I tend to think this series might impose a regression on certain users.

raphaelsc replied:
Users that rely on periodic majors and have many keyspaces might face a problem. But on the other hand, the more keyspaces you have, the smaller your tables are.

bhalevy replied:
That's a valid point.  https://github.com/scylladb/scylladb/pull/15971 will limit the "window of vulnerability" also for major compaction, so users that want to purge every last tombstone in the table should run repair and flush first before running major compaction, but in the common case major compaction can do best efforst with the on-going limitations of tombstone_gc (in whatever mode it is), and now commitlog considerations.

bhalevy commented:
In v2 ([c0a67c6](https://github.com/scylladb/scylladb/commit/c0a67c67ca9e6d7392b846c7e8ff184ddf20f678)):
- `database::flush_all_tables` flushes all tables on that shard (in contrast to `flush_on_all_shards` in previous version)
- Added a `do_flush` flag to `compact_all_sstables` so it doesn't flush again after all tables were already flushed
- Call to `database::flush_all_tables` moved down to `shard_major_keyspace_compaction_task_impl`

bhalevy commented:
@elcallio what's the status of https://github.com/scylladb/scylladb/pull/15971?
Can we move it of RFC mode and merge it?
I think that with it, this PR can be retracted.

https://github.com/scylladb/scylla-dtest/pull/3689 can be merged in any case to stop the annoying test failures.
IIUC, it will be needed also with https://github.com/scylladb/scylladb/pull/15971

elcallio commented:
@bhalevy - needs a consensus on default values + global option for QA etc test purposes.

elcallio commented:
Also, while https://github.com/scylladb/scylladb/pull/15971 helps the real-world case, tests will probably not be fixed by it, since the expiry time (disabled by default so far) is/probably will be higher than the test(s) elect to wait. The stuff done here is still relevant, albeit perhaps requiring a think-through again. 

You could probably reduce flush amount both by figuring out which CF:s will be processed in a compaction, and by checking iff said CF:s have potentially lingering data in the CL in the first place, then figuring which CF:s are holding it up (info is there). Potentially, this could reduce the amount of work done.

bhalevy commented:
> Also, while #15971 helps the real-world case, tests will probably not be fixed by it, since the expiry time (disabled by default so far) is/probably will be higher than the test(s) elect to wait. The stuff done here is still relevant, albeit perhaps requiring a think-through again.
> 
> You could probably reduce flush amount both by figuring out which CF:s will be processed in a compaction, and by checking iff said CF:s have potentially lingering data in the CL in the first place, then figuring which CF:s are holding it up (info is there). Potentially, this could reduce the amount of work done.

It's ok if the tests run `nodetool flush` before major compaction.
As for regular compaction and tombstone purging, it's hard to test that.  Such tests should take into account the config option.

avikivity commented on compaction/task_manager_module.cc:
```diff
@@ -289,6 +289,9 @@ future<> shard_major_keyspace_compaction_task_impl::run() {
     std::vector<table_tasks_info> table_tasks;
+
+    co_await _db.flush_all_tables();
+
```
Let's protect it with a comment, and with a configuration variable in case it causes a regression and we need to disable it in the field.

avikivity commented on compaction/task_manager_module.cc:
```diff
@@ -300,7 +303,8 @@ future<> table_major_keyspace_compaction_task_impl::run() {
     co_await run_on_table("force_keyspace_compaction", _db, _status.keyspace, _ti, [info] (replica::table& t) {
-        return t.compact_all_sstables(info);
+        // All tables were already flushed in shard_major_keyspace_compaction_task_impl
+        return t.compact_all_sstables(info, replica::table::do_flush::no);
```
What does "table_major_keyspace_compaction_task_impl" mean?

What happens on `nodetool compact`? It's not keyspace specific.

bhalevy replied:
> What does "table_major_keyspace_compaction_task_impl" mean?

The task hierarchy is keyspace (top-level), shard, then table.

> What happens on `nodetool compact`? It's not keyspace specific.

Current JMX looks up all keyspaces and calls the api for each keyspace.
We could changes that by presenting a wildcard api to compact all keyspaces (we can add a flag for user only if it's interesting) and then do the uber-flush from there as an optimization.

avikivity replied:
> > What does "table_major_keyspace_compaction_task_impl" mean?
> 
> The task hierarchy is keyspace (top-level), shard, then table.

How did you decode that from table_major_keyspace_compaction_task_impl?


"keyspace" is between "major" and "compaction".

> 
> > What happens on `nodetool compact`? It's not keyspace specific.
> 
> Currently, JMX looks up all keyspaces and calls the api for each keyspace. We could change that by presenting a wildcard api to compact all keyspaces (we can add a flag for user only if it's interesting) and then do the uber-flush from there as an optimization.

We might want to, with many keyspaces it would generate many flushes.

raphaelsc replied:
I raised this concern above. Maybe we could restrict this system-wide flush before major to once a day. We also don't have control over how the user issues major (similar to repair jobs). There could be an application that calls nodetool compact for every table / keyspace at a time.

bhalevy replied:
> I raised this concern above. Maybe we could restrict this system-wide flush before major to once a day. We also don't have control over how the user issues major (similar to repair jobs). There could be an application that calls nodetool compact for every table / keyspace at a time.

There's a limit to the level of protection the the system needs to provide against anti-patterns.
Given a compact_all_keyspaces api, it can print a warning if it's called too frequently.  That's enough IMO.

bhalevy replied:
> There could be an application that calls nodetool compact for every table / keyspace at a time.

We can add a flag to disable all-tables flush, similar to the `skip flush` flag to the snapshot api.
In this case the user is taking responsibility to call (or not to call) `nodetool flush`.

bhalevy commented:
In [db48223](https://github.com/scylladb/scylladb/commit/db4822309a555611039e657d9d4b652568857d41):
- moved call to `flush_all_tables` into `major_keyspace_compaction_task_impl::run`
- added db::config `compaction_flush_all_tables_before_major` (defaults to `true`)
- added TODO comment for adding a `compact_all_keyspaces` api to be called by scylla-nodetool (Cc @denesb)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4676/
- Duration: 6 hr 6 min
- Builder: i-0b6360bf6dbaf89dd

bhalevy commented:
In [883496a](https://github.com/scylladb/scylladb/commit/883496a55a7b6c8de00ce0e784442599ebd65169):
- rebased
- scylla-nodetool, docs: improve optional params documentation (@annastuchlik please review)
- test_tools: scylla_sstable: use no_autocompaction_context
- test_tools: scylla_sstable: move final flush from table_factory
- cql-pytest: nodetool.flush: support global/keyspace flush
- test_tools: test nodetool flush
- cql-pytest: nodetool.compact: support global/keyspace compact
- test_tools: test nodetool compact
- api: force_keyspace_compaction: add skip_flush option
- compaction_manager: flush_all_tables before major compaction
  - limit flush_all_tables rate to 1/24h by default, added config option
- api: add /storage_service/flush
- api: add /storage_service/compact

@denesb please review. In particualr changes to scylla-nodetool and test_tools

bhalevy commented:
In [4a4cfaf](https://github.com/scylladb/scylladb/commit/4a4cfaf48530cdc2f747f3011239849808836bb6):
- fixed test/nodetool issues

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4725/
- Duration: 8 hr 3 min
- Builder: i-0d0a6c2a1eaab2635

denesb commented on test/cql-pytest/test_tools.py:
```diff
@@ -144,10 +144,10 @@ def scylla_sstable(table_factory, cql, ks, data_dir):
-
     try:
-        yield (schema_file, sstables)
+        with nodetool.no_autocompaction_context(cql, f"{ks}.{table}"):
```
All tables in this test already use `NullCompactionStrategy` AFAIK. Did you see any problems here?
Using `no_autocompaction_context` here is not incorrect so I won't protest.

bhalevy replied:
I thought I saw problems, but may I misdiagnozed it (based on the `NullCompactionStrategy` - that I may have proposed myself in the past now that I'm reminded of it)

bhalevy replied:
dropped this patch from the series

denesb commented on test/cql-pytest/test_tools.py:
```diff
@@ -146,6 +136,7 @@ def scylla_sstable(table_factory, cql, ks, data_dir):
 
     try:
         with nodetool.no_autocompaction_context(cql, f"{ks}.{table}"):
+            nodetool.flush(cql, f"{ks}.{table}")
```
Note that all table factories also do intermediate flushes, not just a final flush. They are all designed to produce a certain number of sstables. Taking this control away from them, even if just partially, is weakening the test.

If you want to test a new kind of flush, I suggest to write a new test for it, instead of interfering with existing ones.

bhalevy replied:
dropped all patches test test_tools.py from the series (we can rename it in a followup PR)

denesb commented on test/cql-pytest/test_tools.py:
```diff
@@ -834,3 +834,34 @@ def test_scrub_validate_mode(scylla_path, scrub_workdir, scrub_schema_file, scru
+        assert sstables != []
+
+    with scylla_sstable(simple_clustering_table, cql, test_keyspace, scylla_data_dir, lambda cql, ks, _: flush_with_scylla_nodetool(cql, scylla_path, ks)) as (schema_file, sstables):
+        assert sstables != []
```
These tests don't really belong here. We have `tests/nodetool` for front-end tests and nodetool tests in dtest.
This test-file is for testing scylla-sstable.

bhalevy replied:
ok, it's not quite clear that `test_tools.py` is testing only scylla-sstable.
maybe it'd be better to rename it accordingly.

denesb replied:
Yes, renaming is probably a good idea, to avoid confusion.

bhalevy replied:
let's rename in a followup PR

denesb commented on api/api-doc/column_family.json:
```diff
@@ -84,6 +84,14 @@
                      "paramType":"path"
                   },
+                  {
+                     "name":"sf",
```
please use a non-abbreviated name

bhalevy replied:
I chose it by following the snapshot api example so to stick to the existing convention.
See https://opensource.docs.scylladb.com/stable/operating-scylla/nodetool-commands/snapshot.html
I think that the readability is more important in the tool level and less at the api level that is less interactive.
Ideally, we could support both short and long aliases for the options also at the api parsing level.

bhalevy replied:
In the next version I kept the api option as `cf` is in next version for consistency reasons with the snapshot api.
scylla-nodetool will take only the long option name: `--skip-flush`

denesb replied:
Ok.

avikivity replied:
I don't understand the consistency argument. Someone picked a bad name for a related CLI all, so we must repeat it?

Also, let's avoid the negative and call it flush-memtables and default it to true.

bhalevy replied:
ok, I'l change the name and use positive semantics.

denesb commented on api/api-doc/storage_service.json:
```diff
@@ -728,6 +728,14 @@
                      "paramType":"query"
+                  },
+                  {
+                     "name":"sf",
```
ditto

bhalevy replied:
see above

denesb commented on tools/scylla-nodetool.cc:
```diff
@@ -735,6 +739,8 @@ command-line arguments, the compaction will run on these tables.
 Fore more information, see: https://opensource.docs.scylladb.com/stable/operating-scylla/nodetool-commands/compact.html
 )",
                 {
+                    typed_option<>("skip-flush,sf", "skip flushing tables before major compaction"),
```
Note that boost doesn't support multi-char short-form parameter names. I think this will have to be `--sf` on the command line. I think it is better to have a proper short form (invokable with `-`) or no short form at all.

denesb replied:
Missing: update test in `test/nodetool`.

bhalevy replied:
added tests in `test/nodetool/test_compact.py` in next version

bhalevy replied:
option supports only `"skip-flush"` in next version

denesb commented on compaction/task_manager_module.hh:
```diff
@@ -48,23 +50,36 @@ protected:
 class major_compaction_task_impl : public compaction_task_impl {
 public:
+    enum class flush_mode {
+        unset,
```
Is this a meaningful option compared to `skip`? What will the behaviour be when this is `unset`.

bhalevy replied:
This is for backward compatibility with internal paths that generate a `major_compaction_task_impl`.
I think it would be safe to have an optional flush_mode there and to set it by default to `compacted_tables` which is the preexisting behavior.

bhalevy replied:
`unset` is removed in next version.
the base class sets the default to `compacted_tables` (though it's unused when it's constructure as a base class of `major_compaction_task_executor`)

denesb commented on compaction/task_manager_module.hh:
```diff
@@ -48,23 +50,36 @@ protected:
+        unset,
+        skip,
+        compacted_tables,
+        all_tables
```
It is unclear what these latter two options do.

bhalevy replied:
I'll add some commentary then.

bhalevy replied:
comments added in next version

denesb commented on compaction/task_manager_module.cc:
```diff
@@ -299,6 +309,7 @@ future<> shard_major_keyspace_compaction_task_impl::run() {
 future<> table_major_keyspace_compaction_task_impl::run() {
     co_await wait_for_your_turn(_cv, _current_task, _status.id);
     tasks::task_info info{_status.id, _status.shard};
+    replica::table::do_flush flush(_flush_mode == flush_mode::compacted_tables);
```
What about `all_tables`?

bhalevy replied:
`all_tables` is handled bu global/keyspace tasks.
In case it was set and the decided to skip it since the previous flush_all_tables happened not long enough ago, they convert it downstream to `compacted_tables`, see above:
```
flush_mode fm = _flush_mode == flush_mode::skip || flushed_all_tables ? flush_mode::skip : flush_mode::compacted_tables;
```

bhalevy replied:
We can also change it here to `_flush_mode != flush_mode::skip` for simplicity.

bhalevy replied:
And then the above would look like `flush_mode fm = flushed_all_tables ? flush_mode::skip : _flush_mode;

bhalevy replied:
done the above in next version.

denesb commented on tools/scylla-nodetool.cc:
```diff
@@ -216,9 +216,7 @@ void compact_operation(scylla_rest_client& client, const bpo::variables_map& vm)
-        for (const auto& keyspace : get_keyspaces(client)) {
-            client.post(format("/storage_service/keyspace_compaction/{}", keyspace), std::move(params));
-        }
+        client.post("/storage_service/compact", std::move(params));
```
There is no params here. In fact, we moved an empty params map on each iteration above, which is a bug, albeit a harmless one.

bhalevy replied:
Hmm, please check again.
1. params may have a `sf` flag.
2. which iteration above?  This is in deleted lines of code...

denesb replied:
Yes, I meant the deleted code. I missed that you added the `sf` flag, then indeed this is right (passing empty map is also technically correct anyway).

bhalevy replied:
fixed the loop in next version (the comment is valid for `api: force_keyspace_compaction: add skip_flush option` patch)

denesb commented on tools/scylla-nodetool.cc:
```diff
@@ -402,7 +408,11 @@ void enablegossip_operation(scylla_rest_client& client, const bpo::variables_map
 
 void flush_operation(scylla_rest_client& client, const bpo::variables_map& vm) {
-    const auto [keyspace, tables] = parse_keyspace_and_tables(client, vm);
+    const auto [keyspace, tables] = parse_keyspace_and_tables(client, vm, false);
```
I wanted to prevent such a param to `parse_keyspace_and_tables()`, because the caller can already easily check `vm.count("keyspace")` itself. No need to add noop code to the above method.

bhalevy replied:
So if the keyspace option is required the other caller should verify that?

bhalevy replied:
Since the `keyspace_required` flag is set to `true` by default, existing call sites need not change.

bhalevy replied:
kept in next version.
let's resume discussion after I push it.

denesb replied:
Ok, I have no strong opinion on this, it is fine this way too.

denesb commented on compaction/task_manager_module.cc:
```diff
@@ -297,19 +379,40 @@ static future<bool> maybe_flush_all_tables(sharded<replica::database>& db) {
+        const auto& ks_name = t->schema()->ks_name();
+        const auto& table_name = t->schema()->cf_name();
+        auto it = tables_by_keyspace.try_emplace(ks_name, std::vector<table_info>()).first;
+        it->second.emplace_back(table_name, table_id);
```
`tables_by_keyspce[ks_name].emplace_back(table_name, table_id);`

bhalevy replied:
done in next version

denesb commented on compaction/task_manager_module.cc:
```diff
@@ -258,6 +258,88 @@ future<> run_table_tasks(replica::database& db, std::vector<table_tasks_info> ta
+            // Task responsible for the smallest keyspace.
+            current_task = keyspace_tasks.back().task;
+            keyspace_tasks.pop_back();
+            cv.broadcast();
```
Where is this waited on?

bhalevy replied:
Good point, we need to pass an optional pointer to the condition variable to major_keyspace_compaction_task_impl, and if present, we'll `co_await wait_for_your_turn(_cv, _current_task, _status.id);` there, similar to `table_major_keyspace_compaction_task_impl` with the table tasks.

bhalevy replied:
passing pointers to `cv` and `current_task` to `major_keyspace_compaction_task_impl` in next version.
they are both null when it is invoked directly from the `force_keyspace_compaction` api.

denesb commented on compaction/task_manager_module.cc:
```diff
@@ -258,6 +258,88 @@ future<> run_table_tasks(replica::database& db, std::vector<table_tasks_info> ta
+            current_task = keyspace_tasks.back().task;
+            keyspace_tasks.pop_back();
+            cv.broadcast();
+            co_await current_task->done();
```
AFAIU all tasks are already running and we are just choosing the order in which to wait on them here. I don't understand why that is interesting?

bhalevy replied:
@Deexie may know better.
I copied the code from the existing `run_table_tasks` function implementation

Deexie replied:
All tasks are running, but they are waiting on condition variable. Here we choose the smallest one of the existing, wake its task and wait for it to finish.

denesb replied:
Two questions:
* Won't `cv.broadcast()` wake up all tasks?
* If no, how do we know that `current_task` will be woken up and not some other task?

Deexie replied:
It will, but `wait_for_your_turn` chooses the right task
```
future<> wait_for_your_turn(seastar::condition_variable& cv, tasks::task_manager::task_ptr& current_task, tasks::task_id id) {
    co_await cv.wait([&] {
        return current_task && current_task->id() == id;
    });
}
```

denesb reviewed (changes requested):
> However, compaction is unaware of mutations in the commitlog
and due to sharing commitlog segments by different tables,
it is possible the data or tombstones will get resurrected
by commitlog replay in case the node restarts right after
the respective data and tombstones are purged by (major) compaction
after https://github.com/scylladb/scylladb/commit/f42eb4d1ce321e93b934ce23e128af29d23d66f8.

I don't understand this paragraph. I thought the entire point of https://github.com/scylladb/scylladb/commit/f42eb4d1ce321e93b934ce23e128af29d23d66f8 was to prevent such data resurrections in the first place. This paragraph seems to suggest it is causing it. What gives?

I also don't like that you added tests for the new API endpoint/nodetool to `test/cql-pytest/test_tools.py`. We have `tests/nodetool`, all frontend tests for `scyla-nodetool` should go there. We have nodetool dtests, all end-to-end nodetool tests should go there. We have tests for the API, API tests should go there. Finally, tests for the `nodetool.py` utility in `cql-pytest` should go to `nodetool.py`. Fragmenting tests like this, will just lead to confusion and it will make it impossible to gauge the coverage of a certain tool/module because tests for it are scattered in random places.

bhalevy commented:
> I don't understand this paragraph. I thought the entire point of https://github.com/scylladb/scylladb/commit/f42eb4d1ce321e93b934ce23e128af29d23d66f8 was to prevent such data resurrections in the first place. This paragraph seems to suggest it is causing it. What gives?

Sorry, I was wrong in my assumptions when writing this.  Will rewrite it.

bhalevy commented:
> I also don't like that you added tests for the new API endpoint/nodetool to `test/cql-pytest/test_tools.py`. We have `tests/nodetool`, all frontend tests for `scyla-nodetool` should go there. We have nodetool dtests, all end-to-end nodetool tests should go there. We have tests for the API, API tests should go there. Finally, tests for the `nodetool.py` utility in `cql-pytest` should go to `nodetool.py`. Fragmenting tests like this, will just lead to confusion and it will make it impossible to gauge the coverage of a certain tool/module because tests for it are scattered in random places.

Okay, I'll rearrange the newly added tests respectively.
So what is `cql-pytest/test_tools.py` for?

bhalevy commented on test/cql-pytest/nodetool.py:
```diff
@@ -71,8 +71,17 @@ def run_nodetool(cql, *args):
+    if not '.' in ks_or_table:
+        flush_keyspace(cql, ks_or_table)
+        return
+    ks, cf = ks_or_table.split('.')
```
patch dropped in next version

bhalevy commented on test/cql-pytest/nodetool.py:
```diff
@@ -93,13 +93,25 @@ def flush_keyspace(cql, ks):
+        requests.post(f'{rest_api_url(cql)}/storage_service/keyspace_compaction/{ks}')
+    else:
+        run_nodetool(cql, "compact", ks)
+
```
patch dropped in next version

bhalevy commented on test/cql-pytest/test_tools.py:
```diff
@@ -865,3 +867,35 @@ def test_flush_keyspace(cql, test_keyspace, scylla_path, scylla_data_dir):
+        assert len(sstables) == SMP
+
+    with scylla_sstable(simple_clustering_table, cql, test_keyspace, scylla_data_dir, lambda cql, ks, _: compact_with_scylla_nodetool(cql, scylla_path, ks)) as (schema_file, sstables):
+        assert len(sstables) == SMP
```
patch dropped in next version

bhalevy commented:
In v4 ([84d1717](https://github.com/scylladb/scylladb/commit/84d1717e91e2489075296570c191710c694b7e46))
- dropped all changes to `test/cql-pytest/{nodetool,test_tools}.py`
- revised the commitlog comments regarding the need for flushing all tables and the 24h interval
- cleaned up the rest based on @denesb excellent review comments.

denesb commented:
> revised the commitlog comments regarding the need for flushing all tables and the 24h interval

Please also update the cover-letter.

denesb commented on test/nodetool/test_flush.py:
```diff
@@ -8,7 +8,13 @@
 
 
-def test_flush(nodetool):
+def test_flush_all_tables(nodetool):
```
You probably need to add `scylla_only` to this test, as the Java nodetool will not use this new endpoint.

Please make sure to run these test against the Java nodetool as well, and adjust the new tests as needed.
You can do that by e.g.:
```
cd tests/nodetool
pytest --nodetool=cassandra test_flush.py
```

bhalevy replied:
done in v5

denesb reviewed (commented):
Looks good, minor comments.

Please also run the new tests in `tests/nodetool` against the Java nodetool:
```
cd tests/nodetool
pytest --nodetool=cassandra test_file.py
```

Please update the cover-letter.

@Deexie please review the task-manager parts.

bhalevy commented:
In v5 ([4fe14df](https://github.com/scylladb/scylladb/commit/4fe14dfdc57f9adfa0c67abc7221c49eef5fc9aa)):
- test/nodetool: jmx: fix path to scripts/scylla-jmx
- added `cassandra_only` and `scylla_only` tests in `test/nodetool/test_{compact,flush}.py`

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4759/
- Duration: 2 hr 50 min
- Builder: spider2.cloudius-systems.com

avikivity commented:
> Major compaction already flushes each table to make sure it considers any mutations that are present in the memtable for the purpose of tombstone purging. See [64ec1c6](https://github.com/scylladb/scylladb/commit/64ec1c6ec64291064678e3818e3ffdee4294dbfe)
> 


Isn't the subject inaccurate then? The series is about flushing the commitlog only (and flushes not-directly-referenced tables as a means to achieve that).

avikivity commented on tools/scylla-nodetool.cc:
```diff
@@ -735,6 +739,8 @@ command-line arguments, the compaction will run on these tables.
 Fore more information, see: https://opensource.docs.scylladb.com/stable/operating-scylla/nodetool-commands/compact.html
 )",
                 {
+                    typed_option<>("skip-flush", "skip flushing tables before major compaction"),
```
Here, too, we should use the positive sense.

avikivity replied:
and maybe explain the implications a bit more

avikivity replied:
ok, this is a switch, not a boolean, so can't reverse it

bhalevy replied:
right

bhalevy replied:
but it looks like we can use typed_option<bool> which defaults to `true`

avikivity commented on compaction/task_manager_module.cc:
```diff
@@ -281,14 +282,38 @@ sstring major_compaction_task_impl::to_string(flush_mode fm) {
+static future<bool> maybe_flush_all_tables(sharded<replica::database>& db) {
+    auto interval = db.local().get_config().compaction_flush_all_tables_before_major_seconds();
+    if (interval) {
+        auto when = db_clock::now() - std::chrono::duration_cast<db_clock::duration>(std::chrono::seconds(interval));
```
I'm guessing you can write `auto when = db_clock::now() - interval * 1s;`

bhalevy replied:
done in v6

bhalevy commented:
> > Major compaction already flushes each table to make sure it considers any mutations that are present in the memtable for the purpose of tombstone purging. See [64ec1c6](https://github.com/scylladb/scylladb/commit/64ec1c6ec64291064678e3818e3ffdee4294dbfe)
> 
> Isn't the subject inaccurate then? The series is about flushing the commitlog only (and flushes not-directly-referenced tables as a means to achieve that).

Fair point.

bhalevy commented:
In v6 ([1bf4ab4](https://github.com/scylladb/scylladb/commit/1bf4ab4aeed72d808e68bc6dd7d8790ae8c54857)):
- updated PR title
- changed compaction api option to `flush_memtables`, true by default
  - modernized the api param handling using `req_params` that rejects unrecognized options (Ref #10087)
- correspondingly, changed `scylla-nodetool compact` option to `--flush-memtables`, true by default
- updated tests to test both `true` and `false` args
- added test for the `column_family/major_compaction` api

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4761/
- Duration: 2 hr 45 min
- Builder: spider5.cloudius-systems.com

denesb commented on test/nodetool/conftest.py:
```diff
@@ -68,7 +68,7 @@ def jmx(request, rest_api_mock_server):
     jmx_path = request.config.getoption("jmx_path")
     if jmx_path is None:
-        jmx_path = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", "..", "scylla-jmx", "scripts",
+        jmx_path = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", "tools", "jmx", "scripts",
```
Right, I think this happened to work on my box because I have such a directory next to `scylla.git`.

Deexie commented on compaction/task_manager_module.hh:
```diff
@@ -76,8 +91,10 @@ public:
+            std::vector<table_info> table_infos,
+            std::optional<flush_mode> fm = std::nullopt) noexcept
+        : major_compaction_task_impl(module, tasks::task_id::create_random_id(), module->new_sequence_number(), "keyspace", std::move(keyspace), "", "", tasks::task_id::create_null_id(),
+                fm.value_or(flush_mode::all_tables))
```
Why does `flush_mode` value set here differ from `major_compaction_task_impl`'s default?

Deexie replied:
Oh, nevermind, I see it's used in further commit

nyh commented on api/api-doc/column_family.json:
```diff
@@ -84,6 +84,14 @@
                   },
+                  {
+                     "name":"flush_memtables",
+                     "description":"Controls flushing of memtables before compaction (true by default). Set to \"false\" to skip automatic flushing of memtables before compaction, e.g. when the table is flushed explicitly invoking the compaction api.",
```
I don't understand this explanation. What do you mean "e.g., when the table is flushed explicitly invoking the compaction api"? Did you forget the word **before**? I.e., "when the table was flushed explicitly **before** invoking the compacation api"?

In general, I don't like the phenomenon that recently we're adding too many obscure options to the REST API which solve in a manual way some very specific problems that nobody will remember in a year (or tomorrow) that even need to be solved. If these are important as manual options, they should be clearly documented in a document - not just one line. If they are not important as manual options, they should have been automatic.

nyh replied:
I see below that you have another copy of this JSON, and it has exactly the word "before" I was missing.
Please choose the best text for both.

bhalevy replied:
sure, will add "before".
The option is not needed if `scylla nodetool compact` is used, then everything is done automatically and efficiently.

bhalevy replied:
fixed in v7

nyh commented on test/nodetool/test_compact.py:
```diff
@@ -76,3 +77,22 @@ def test_user_defined(nodetool, scylla_only):
+
+
+@pytest.mark.parametrize("flush", ("true", "false"))
+def test_keyspace_flush(nodetool, scylla_only, flush):
```
It's not clear what a "keyspace flush" actually test? It doesn't test "nodetool flush"... Please add a comment explaining that it actually tests a specific parameter of "nodetool compact". Maybe also use "compact" in the test's name.
And please also explain why it's a Scylla-only test, because it tests a nodetool option "--flush-memtables" that didn't exist in Cassandra's nodetool.

bhalevy replied:
> It's not clear what a "keyspace flush" actually test? It doesn't test "nodetool flush"... Please add a comment explaining that it actually tests a specific parameter of "nodetool compact".

I'll rename the function to mention `flush_memtables_option`.

> Maybe also use "compact" in the test's name.

test_compaction.py is all about `scylla nodetool compact` and it's not mentioned in any other test case.
I think that given the mention of the specific option in the function name, this isn't requried.

> And please also explain why it's a Scylla-only test, because it tests a nodetool option "--flush-memtables" that didn't exist in Cassandra's nodetool.

No problem.

bhalevy replied:
done in v7

nyh commented on db/config.cc:
```diff
@@ -361,6 +361,10 @@ db::config::config(std::shared_ptr<db::extensions> exts)
+    , compaction_flush_all_tables_before_major_seconds(this, "compaction_flush_all_tables_before_major_seconds", value_status::Used, 86400,
+        "Set the minimum interval in seconds between flushing all tables before each major compaction (default is 86400). "
+        "This option is useful for maximizing tombstone garbage collection by releasing all active commitlog segments. "
+        "Set to 0 to disable automatic flushing all tables before major compaction")
```
I didn't understand this description. The default is 86400 seconds, i.e., a day. What does it mean? If I do a "nodetool compact" it will or won't flush the tables? What does "a day" do here - does it mean that if the previous nodetool compact happened more than a day or less than a day, the behavior is different?

bhalevy replied:
Yes, flush of **all** tables happens at most once a day by default since it is expensive and may create small sstables due to excessive memtable flushing.
The reason is to deal with the user / management tool invoking the `keyspace_compaction` api repeatedly, for each keyspace. It's sufficient to flush all tables only once, on the first invocation.
If more than a day passed since last flush then all tables would be flush again.

BTW, this is related to https://github.com/scylladb/scylladb/pull/15971 that will help flush idle memtable in a timely manner, and with it this mechanism here should skip tables that were already flushed in the last 24 hours, but we're not yet committed to it.

nyh commented on test/rest_api/test_column_family.py:
```diff
@@ -85,15 +85,23 @@ def test_column_family_tombstone_gc_correctness(cql, this_dc, rest_api):
-            # 4 Flush it into sstable
-            resp = rest_api.send("POST", f"storage_service/keyspace_flush/{test_keyspace}")
+            # 4 Trigger major compaction on that sstable (flushes memtable by default)
+            resp = rest_api.send("POST", f"column_family/major_compaction/{test_keyspace}:{test_table}")
```
I don't understand something here.... It seems you are assuming that major compaction will flush the memtable. But I see two patches later that there is some 1-day threshold that may avoid this flush. So is this flush guaranteed? Was it always guaranteed, also before this patch, and the flush step was always unnecessary?

bhalevy replied:
The addition in this series is the flush of **all** tables in the database before major compaction.
Flushing the particular tables that are major compacted was already done before (in all supported versions), and is still done in the case that flushing all db tables is skipped due to the 1-day threshold (but not if the `flush_memtables=false` option is given).

nyh commented on api/api-doc/storage_service.json:
```diff
@@ -920,6 +920,21 @@
+         "operations":[
+            {
+               "method":"POST",
+               "summary":"Flush all memtables in all keyspaces.",
```
It is sad that the real explanation why this API exists is hidden in a commit message and not visible to the users. We need to start investing more efforts into documentation (for users) and comments (for developers), than in commit messages :-(

bhalevy replied:
The functionality exists forever at the `nodetool` level, the new api is there to allow nodetool/jmx to implement it in a more efficient way.  The api level is not exposed to the end user, but rather to field-engineering and to power-admins.  The `summary` text here can provide some additional information about the call usage, but I'm not sure the motivation for it is necessarily required here.

nyh commented on test/nodetool/conftest.py:
```diff
@@ -141,6 +141,12 @@ def scylla_only(request):
+@pytest.fixture(scope="function")
+def cassandra_only(request):
+    if request.config.getoption("nodetool") != "cassandra":
+        pytest.skip('Cassandra-only test skipped')
```
Ideally, such a fixture should not exist:
1. If this is a cassandra feature that **should** be in Scylla, it's a Scylla bug, and it should be marked "@xfail", not "cassandra_only".
2. If this is a cassandra feature that should never be a Scylla, we shouldn't have a test for it - we're not writing a test suite for Cassandra.

nyh replied:
Ok, I understand better now why you added cassandra_only - because you wanted the Java ("cassandra") nodetool to behave differently than the C++ one. More on this below.

bhalevy replied:
yes, it's not a cassandra-only feature, but rather cassandra-specific implementation detail that is tested (in a grey-box way)

nyh commented on test/nodetool/test_flush.py:
```diff
@@ -8,7 +8,22 @@
+def test_flush_all_tables(nodetool, scylla_only):
+    nodetool("flush", expected_requests=[
+        expected_request("POST", "/storage_service/flush")
+    ])
```
Something very strange is happening here, and I'd like @denesb 's opinion.
You decided to change the behavior of "nodetool flush" (to call some new Scylla-only API), but decided to do this only in the new nodetool command that nobody uses yet - and not in the Java nodetool. And these tests verify that this is the case. But why? Why should the C++ nodetool behave differently from the Java one? The whole point of this test file is to verify that it doesn't. At least not yet (I'm sure that in the future we'll stop adding features to the old nodetool, but have we reached that state?)

bhalevy replied:
@nyh it doesn't change the "black-box" behavior from the user's perspective.
It achieves the same goal for bare `nodetool flush` in either `nodetool` implementations.
It is a mistake to have scylla_nodetool just mimic the exact api calls that happen to happen in the json nodetool implementation over scylla-jmx, as it inhibits progress.
If the semantics of the operation are indistinguishable I see no reason not to evolve the underlying implementation.

denesb replied:
I agree, as long as there are no semantical differences, the two implementations can diverge.
For the most part, there is no reason to diverge, but when there are more optimal set of API endpoints available, there is no reason to not use it.
I suspect we will see more and more changes like this, and slowly, even semantic differences will creep in. The reason is that updating the Java nodetool is such a chore, that everybody avoids it as much as they can.

nyh commented on api/api-doc/storage_service.json:
```diff
@@ -701,6 +701,30 @@
+         "operations":[
+            {
+               "method":"POST",
+               "summary":"Forces major compaction in all keyspaces",
```
To be honest, I don't understand when this API is ever useful. Maybe just when you have exactly one big table?
I remember in the past we tried to claim that users should never explicitly call major compaction at all - that it just ruins performance. So now we're giving them a way to do a major compaction of all tables at once?

bhalevy replied:
In dtest (that represents users mindset) bare `nodetool compact`, that causes major compaction of all keyspaces, is used elaborately, and naive users may also invoke it sparringly.
Unfortunately in many use cases major compaction is still required once in a while to prevent accumulation of deleted data and tombstones, though in this case I admit the usage may typically be needed for a specific table or keyspace.
In any case, the API is there also to help field-engineering in extreme corner cases.  Having an API for something doesn't mean we're encouraging anyone to use it on a regular basis.  It's a power-tool that should be used with care, understanding what one is doing.

bhalevy commented on compaction/task_manager_module.hh:
```diff
@@ -82,20 +82,44 @@ protected:
+            std::optional<flush_mode> fm = std::nullopt,
+            seastar::condition_variable* cv = nullptr,
+            tasks::task_manager::task_ptr* current_task = nullptr) noexcept
+        : major_compaction_task_impl(module, tasks::task_id::create_random_id(), module->new_sequence_number(), "keyspace", std::move(keyspace), "", "", parent_id,
```
fixed in v7

bhalevy commented:
In v7 ([b5e2603](https://github.com/scylladb/scylladb/commit/b5e26036b4259fd63b75fbb131e960af73a4de25)):
- rebased
- fixed task sequence number generation of keyspace_major_compaction_task_impl (see https://github.com/scylladb/scylladb/pull/15820#discussion_r1399131713)
- Improved flush_memtables option description phrasing in api (to include missing "before", see https://github.com/scylladb/scylladb/pull/15820#discussion_r1399750982)
- Used `_flush_memtables_option` in test function names in `test/nodetool/test_compact.py` (see https://github.com/scylladb/scylladb/pull/15820#discussion_r1399755900)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4829/
- Duration: 2 hr 28 min
- Builder: spider7.cloudius-systems.com

bhalevy commented:
rebased

bhalevy commented:
In v8 ([310ff20](https://github.com/scylladb/scylladb/commit/310ff20e1e2ea49024a6bf00abd177444e7972a8)):
- rebased + retested
- docs: nodetool: compact: fix example
- docs: nodetool: flush: enrich examples

@annastuchlik can you please review last 2 patches.
Sorry for the last minute changes.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4944/
- Duration: 2 hr 23 min
- Builder: spider1.cloudius-systems.com

bhalevy commented:
@denesb / @scylladb/scylla-maint please queue
